[#49527] Allow Shared with users to become assignees

aaron-contreras opened the pull request:
See: https://community.openproject.org/work_packages/49527

aaron-contreras committed:
Performance: Improve possible assignee scope spec

Make good use of shared lets :)

aaron-contreras committed:
Add code path to ensure only legal arguments are provided

Individual WorkPackages, Projects and a collection of unique resources
of either are considered legal arguments. Anything else will raise
and ArgumentError.

aaron-contreras committed:
Ensure active users, members of a work package are returned

aaron-contreras committed:
Ensure member placeholder users on a work package are returned

aaron-contreras committed:
Ensure member groups on a work package are returned

aaron-contreras committed:
Ensure invited users are returned as well

aaron-contreras committed:
Ensure locked users are not returned

aaron-contreras committed:
Ensure only users with an assignable role are returned

aaron-contreras committed:
Ensure multiple work package arguments return the intersection of possible assignees

aaron-contreras committed:
Extract work package query helper method

aaron-contreras committed:
Extract project query helper method

aaron-contreras committed:
Add OAS3 doc for new available_assignees endpoint

aaron-contreras committed:
Register /work_packages/{id}/available_assignees API

* Generates API class

* Mounts API to work_packages_api

aaron-contreras committed:
Include Project members with a valid role in the work package possible assignee scope

aaron-contreras committed:
Update API path helpers and contract

aaron-contreras committed:
Ensure user count is scope specific

aaron-contreras committed:
Adapt specs to new behavior

aaron-contreras committed:
Merge branch 'dev' into implementation/49527-add-work-package-assignable-scopes-and-api

aaron-contreras committed:
Allow all subclasses of WorkPackages to be assignable

aaron-contreras commented on spec/support/api/v3/shared_available_principals_examples.rb:
```diff
@@ -104,6 +114,11 @@
 
-    it_behaves_like 'unauthorized access'
+    if work_package_scope
+      it_behaves_like 'not found',
+                      I18n.t('api_v3.errors.not_found.work_package')
```
Default behavior for work-package-scoped endpoints is not to reveal the presence of restricted work packages

ulferts commented on lib/api/v3/work_packages/available_assignees_api.rb:
```diff
@@ -0,0 +1,50 @@
+      class AvailableAssigneesAPI < ::API::OpenProjectAPI
+        resource :available_assignees do
+          after_validation do
+            authorize_in_work_package(:view_work_packages, work_package: @work_package)
```
Why is the permission asked for here `:view_work_packages`? I realise that it is the same permission queried for by the `Projects::AvailableAssigneesAPI` as well. But a user being allowed to see the work packages does not need to see every assignable user. Only users having the `:edit_work_packages` permission need to see it here. For the `Projects::AvailableAssigneesAPI` endpoint, the permission should be `:add_work_packages` by now I think as that endpoint will only be used when creating a new work package now.

aaron-contreras replied:
You're right. Nice catch!

ulferts commented on docs/api/apiv3/paths/work_package_available_assignees.yml:
```diff
@@ -0,0 +1,114 @@
+              value:
+                _type: Error
+                errorIdentifier: urn:openproject-org:api:v3:errors:NotFound
+                message: The specified work package does not exist.
```
The message actually seems to be ` The work package you are looking for cannot be found or has been deleted.`.

ulferts commented on docs/api/apiv3/paths/work_package_available_assignees.yml:
```diff
@@ -0,0 +1,114 @@
+        **Required permission:** view work packages
+
+        *Note that you will only receive this error, if you are at least allowed to see the corresponding work package.*
+      headers: {}
```
With the way the API is currently structured, this case cannot happen. If the user lacks the `view work packages` permission, they will not see the work package (`/api/v3/work_packages/:id`) which is the precondition to reaching its assignees (`/api/v3/work_packages/:id/available_assignees`).

This might change if the permission is changed to `edit_work_packages`

ulferts reviewed (changes requested):
The change is looking good @aaron-contreras with the exception of a permission check which I think can be tightened.

aaron-contreras committed:
Correct permissions checked in order to view available assignees

aaron-contreras committed:
Update OAS doc for project_available_assignees.yml

* Sets the right permission requirement in the documentation.

aaron-contreras committed:
Update OAS doc for work_package_available_assignees.yml

* Updates the required permissions and the documented error response.

aaron-contreras committed:
Extract base permission set into shared example caller

This cleans up the readability of the spec given it's used in multiple
contexts. Easier to parse when you can refer directly from the caller
you're trying to inspect.

aaron-contreras commented:
Feedback addressed. The specs should read a bit better now as well :) @ulferts

klaustopher reviewed (approved):
lgtm
